src/handlers.rs

```diff
@@ -134,6 +142,125 @@ pub fn handle_request(
             Ok(vec![])
         }
 
+        RequestType::GETAI => {
+            let hostname = CStr::from_bytes_with_nul(request.key)?.to_str()?;
+            // Boths hints are necessary to mimick the glibc behaviour.
+            let hints = AddrInfoHints {
+                // The canonical name will be filled in the first
+                // addrinfo struct returned by getaddrinfo.
+                flags: AI_CANONNAME,
+                // Deduplicate the resulting addrs.
+                socktype: SOCK_STREAM,
+                address: 0,
+                protocol: 0,
+            };
+            let resp = dns_lookup::getaddrinfo(Some(hostname), None, Some(hints));
+            let ai_resp_empty = AiResponse {
+                canon_name: hostname.to_string(),
+                addrs: vec![],
+            };
+            let ai_resp: AiResponse = match resp {
+                Ok(ai_resp_iter) => {
+                    let mut ai_resp_iter = ai_resp_iter.filter_map(|e| e.ok()).peekable();
+                    // According to man 3 getaddrinfo, the resulting
+                    // canonical name should be stored in the first
+                    // addrinfo struct.
+                    // Re-using the request hostname if we don't get a
+                    // canonical name.
+                    let canon_name = ai_resp_iter
+                        .peek()
+                        .and_then(|e| e.canonname.to_owned())
+                        .unwrap_or(hostname.to_string());
+                    let addrs: Vec<IpAddr> = ai_resp_iter
+                            .map(|e| e.sockaddr.ip())
+                            .collect();
+
+                    AiResponse {
+                        canon_name,
+                        addrs,
+                    }
+                }
+                Err(_) => ai_resp_empty,
+            };
+
+            serialize_address_info(&ai_resp)
+        }
+
+        // GETHOSTBYADDR and GETHOSTBYADDRv6 implement reverse lookup
+        // The key contains the address to look for.
+        RequestType::GETHOSTBYADDR => {
+            let key = request.key;
+
+            if key.len() != 4 {
+                bail!("Invalid key len: {}, expected 4", key.len());
+            }
+            let address_bytes: [u8; 4] = key.try_into()?;
+            let hostent = match gethostbyaddr_r(LibcIp::V4(address_bytes)) {
+                Ok(hostent) => hostent,
+                Err(e) =>
+                // We shouldn't end up in that branch. Something
+                // got very very wrong on the glibc client side if
+                // we do. It's okay to bail, there's nothing much
+                // we can do.
+                {
+                    bail!("unexpected gethostbyaddr error: {}", e)
+                }
+            };
```

Comment:
```suggestion
            let hostent = gethostbyaddr_r(LibcIp::V4(address_bytes))?;
```

I probably would have written this, and just let the original error surface and get logged. But I don't have strong (or generally good) opinions about error handling. I'll defer to @blinsay if he wants to have a stronger or better opinion.

(same idea below too)

Reply:
Did not address this one yet. Let me know what you prefer.

Reply:
The whole thing with the sentinel error-case hostent getting returned from `gethostbyaddr_r` is very odd to me. We don't have to reproduce libc's strangeness exactly, even if we do have to serialize that error value and send it back to the caller.

What if we did something like:

```rust
enum HostentError {
  NotFound,
  TryAgain,
  NoRecovery,
  NoData,
  Other(anyhow::Error),
}

pub fn gethostbyname2_r(name: String, af: libc::c_int) -> Result<Hostent, HostentError> {
  ...
}
```

and then in handlers we could do something like this:

```rust
            let hostent = match gethostbyaddr_r(LibcIp::V4(address_bytes)) {
                Ok(hostent) => hostent,
                Err(Other(e)) => bail!("oh no, something very bad happened: {e}"),
                Err(e) => {
                  log::debug!("gethostbyaddr_r(..) = {e:?}");
                  Hostent::error_value()
                },
            };
```